File: bench/cases/prime-sieve/prime-sieve.js
// prime-sieve: Sieve of Eratosthenes up to LIMIT, counts primes.
// Uint8Array for the sieve buffer (idiomatic fast JS). The `p <= limit/p`
// guard keeps p*p within 32-bit range.
"use strict";

const limit = 30000000;
const expected = 1857859;

const sieve = new Uint8Array(limit + 1);
let count = 0;
for (let p = 2; p <= limit; p++) {
  if (sieve[p] === 0) {
    count++;
    if (p <= Math.floor(limit / p)) {
      for (let j = p * p; j <= limit; j += p) sieve[j] = 1;
    }
  }
}
const checksum = count;
if (checksum !== expected) {
  process.stderr.write(`MISMATCH got=${checksum} want=${expected}\n`);
  process.exit(1);
}
process.exit(0);

File: bench/cases/prime-sieve/prime-sieve.go
package main

// prime-sieve: Sieve of Eratosthenes up to LIMIT, counts primes.
// Exercises array allocation, indexing and nested loops.
// The `p <= LIMIT/p` guard keeps p*p within 32-bit range for every language.

import (
	"fmt"
	"os"
)

const limit = 30000000
const expected int64 = 1857859

func main() {
	sieve := make([]byte, limit+1)
	count := int64(0)
	for p := 2; p <= limit; p++ {
		if sieve[p] == 0 {
			count++
			if p <= limit/p {
				for j := p * p; j <= limit; j += p {
					sieve[j] = 1
				}
			}
		}
	}
	checksum := count
	if expected < 0 {
		fmt.Fprintf(os.Stderr, "checksum=%d\n", checksum)
		os.Exit(0)
	}
	if checksum != expected {
		fmt.Fprintf(os.Stderr, "MISMATCH got=%d want=%d\n", checksum, expected)
		os.Exit(1)
	}
	os.Exit(0)
}
